pilot/pkg/proxy/envoy/v2/debug.go

```diff
@@ -319,53 +290,99 @@ func (s *DiscoveryServer) authenticationz(w http.ResponseWriter, req *http.Reque
-					info.TLSConflictStatus = "MAY CONFLICT"
-				} else {
-					info.TLSConflictStatus = "CONFLICT"
-				}
-			} else {
-				info.TLSConflictStatus = "OK"
-			}
-			if b, err := json.MarshalIndent(info, "  ", "  "); err == nil {
-				_, _ = w.Write(b)
-			}
-			_, _ = fmt.Fprintf(w, ",\n")
+	if b, err := json.MarshalIndent(info, "  ", "  "); err == nil {
+		_, _ = w.Write(b)
+	}
+}
+
+// AnalyzeMTLSSettings returns mTLS compatibility status between client and server policies.
+func AnalyzeMTLSSettings(hostname host.Name, port *model.Port, authnPolicy *authn.Policy,
+	destConfig *model.Config) []*AuthenticationDebug {
+	// TODO(diemvu): add policy config name to the cache push config for this.
+	authnPolicyName := "-"
+	if authnPolicy != nil {
+		authnPolicyName = "???"
+	}
+
+	serverMTLSMode := authn_alpha1.GetMutualTLSMode(authnPolicy)
+
+	baseDebugInfo := AuthenticationDebug{
+		Port:                     port.Port,
+		AuthenticationPolicyName: authnPolicyName,
+		DestinationRuleName:      configName(destConfig),
+		ServerProtocol:           serverMTLSMode.String(),
+		ClientProtocol:           "-",
+	}
+
+	var rule *networking.DestinationRule
+
+	if destConfig != nil {
+		rule = destConfig.Spec.(*networking.DestinationRule)
+	}
+
+	output := []*AuthenticationDebug{}
+
+	clientTLSModes := collectTLSSettingsForPort(rule, port)
+	var subsets []string
+	for k := range clientTLSModes {
+		subsets = append(subsets, k)
+	}
+	sort.Strings(subsets)
+
+	for _, ss := range subsets {
+		c := clientTLSModes[ss]
+		info := baseDebugInfo
+		if c != nil {
+			info.ClientProtocol = c.GetMode().String()
 		}
+		if ss == "" {
+			info.Host = string(hostname)
+		} else {
+			info.Host = fmt.Sprintf("%s|%s", hostname, ss)
+		}
+		info.TLSConflictStatus = EvaluateTLSState(c, serverMTLSMode)
+
+		output = append(output, &info)
 	}
-	_, _ = fmt.Fprint(w, "\n{}]")
+	return output
+}
+
+// EvaluateTLSState returns the conflict state (string) for the input client+server settings.
+// The output string could be:
+// - "AUTO": client TLS settings is not set, and will be automatically determined based on server mode.
+// - "OK": both client and server TLS settings are set correctly.
+// - "CONFLICT": both client and server TLS settings are set, but could be incompatible.
+func EvaluateTLSState(clientMode *networking.TLSSettings, serverMode authn_alpha1.MutualTLSMode) string {
+	const okState string = "OK"
+	const conflictState string = "CONFLICT"
+	const autoState string = "AUTO"
+
+	if clientMode == nil {
+		return autoState
+	}
+
+	if (serverMode == authn_alpha1.MTLSDisable && clientMode.GetMode() == networking.TLSSettings_DISABLE) ||
+		(serverMode == authn_alpha1.MTLSStrict && clientMode.GetMode() == networking.TLSSettings_ISTIO_MUTUAL) ||
+		(serverMode == authn_alpha1.MTLSPermissive &&
+			(clientMode.GetMode() == networking.TLSSettings_ISTIO_MUTUAL || clientMode.GetMode() == networking.TLSSettings_DISABLE)) {
+		return okState
+	}
+
+	return conflictState
```

Comment:
An TODO for me or whoever to change this with PERMISSIVE mode once it's implemented and the mesh_config enables it.

Reply:
hmm, nvm, i saw you have that in 375.